File: frontend/src/utils/cyclePhase.ts
import type { CyclePhase } from "@/services/types";

/**
 * Mesma regra usada no backend (offers/cycle_phase.py) e no projeto 1:
 * - "matricula": até o fim do período de matrícula (enrollment_end), inclusive.
 * - "andamento": entre o fim da matrícula e o fim do curso (course_end), inclusive.
 * - "encerramento": após o fim do curso.
 */
export function computeCyclePhase(
  enrollmentEnd: string,
  courseEnd: string,
  reference: Date = new Date()
): CyclePhase {
  const today = toDateOnly(reference);
  const end = toDateOnly(new Date(enrollmentEnd));
  const courseEndDate = toDateOnly(new Date(courseEnd));

  if (today <= end) return "matricula";
  if (today <= courseEndDate) return "andamento";
  return "encerramento";
}

function toDateOnly(date: Date): number {
  return Date.UTC(date.getFullYear(), date.getMonth(), date.getDate());
}

File: backend/offers/cycle_phase.py
"""Regra de cálculo da fase do ciclo de uma oferta.

Mesma regra usada no projeto 1 (vitrine de ofertas):
- "matricula": até o fim do período de matrícula (enrollment_end).
- "andamento": entre o fim da matrícula e o fim do curso (course_end).
- "encerramento": após o fim do curso.
"""

from datetime import date

MATRICULA = "matricula"
ANDAMENTO = "andamento"
ENCERRAMENTO = "encerramento"

CYCLE_PHASE_CHOICES = [
    (MATRICULA, "Matrícula"),
    (ANDAMENTO, "Andamento"),
    (ENCERRAMENTO, "Encerramento"),
]


def compute_cycle_phase(
    enrollment_end: date,
    course_end: date,
    reference: date | None = None,
) -> str:
    """Calcula a fase do ciclo a partir das datas da oferta.

    - Até enrollment_end (inclusive): matrícula.
    - Entre enrollment_end e course_end (inclusive): andamento.
    - Após course_end: encerramento.
    """
    today = reference if reference is not None else date.today()

    if today <= enrollment_end:
        return MATRICULA
    if today <= course_end:
        return ANDAMENTO
    return ENCERRAMENTO
